m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -556,21 +556,27 @@ public void run() {
                 }
             }
 
-            state.set(ProcessorState.SHUTDOWN);
-
             LOG.warn("Processor shutting down");
 
+            if (isShutdown.get()) {
+                // If the processor hit an exception, another processor will be created to resume work on the queue,
+                // so leave the queue as is.
+                drainAndFlush();
+            }
             shutdown();
+            state.set(ProcessorState.SHUTDOWN);
 
             LOG.warn("Processor shut down");
         }
 
-        private void shutdown() {
+        private void drainAndFlush() {
```

Comment:
There's very little sense of breaking this out, instead just move conditional into `shutdown`